### PR TITLE
[ggj] feat: support box/primitive type equality for ternary expression

### PR DESCRIPTION
**Pull request by xiaozhenliu-gg5**
[3/6] support boxed/primitive type equality in ternayexpression.

More details can be found in notes: go/javanaut-boxed-primitive-type-equality

### CODE REVIEW

File: src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java [hunk 1]
```diff
@@ -23,9 +23,6 @@
 public class TernaryExprTest {
   @Test
   public void validTernaryExpr_primitiveType() {
-    Variable variable = Variable.builder().setName("x").setType(TypeNode.INT).build();
-    VariableExpr variableExpr = VariableExpr.builder().setVariable(variable).build();
-
     Variable conditionVariable =
         Variable.builder().setName("condition").setType(TypeNode.BOOLEAN).build();
```
Comment (xiaozhenliu-gg5) [inline src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java:27]: unused variableExpr, same below. The related ternary fix is separated to another PR: #187

File: src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java [hunk 2]
```diff
@@ -71,6 +65,25 @@ public void validTernaryExpr_objectType() {
...
+    Variable elseVariable =
+        Variable.builder().setName("integerValue").setType(TypeNode.INT_OBJECT).build();
+    VariableExpr elseExpr = VariableExpr.builder().setVariable(elseVariable).build();
+    TernaryExpr ternaryExpr =
+        TernaryExpr.builder()
+            .setConditionExpr(conditionExpr)
+            .setThenExpr(thenExpr)
+            .setElseExpr(elseExpr)
+            .build();
+    assertThat(ternaryExpr.conditionExpr().type()).isEqualTo(TypeNode.BOOLEAN);
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java:84]: Please add an assert for `ternaryExpr.type()`.

File: src/main/java/com/google/api/generator/engine/ast/TernaryExpr.java [hunk 3]
```diff
@@ -54,13 +54,19 @@ public static Builder builder() {
...
 
     abstract TernaryExpr autoBuild();
 
     public TernaryExpr build() {
+      Preconditions.checkState(
+          conditionExpr().type().equals(TypeNode.BOOLEAN),
+          "Ternary condition must be a boolean/Boolean-typed expression.");
+
       if (!thenExpr().type().equals(elseExpr().type())) {
         // Not both primitives, and no boxed equality, and one of them is null.
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/TernaryExpr.java:71]: Could we please update this comment?

File: src/main/java/com/google/api/generator/engine/ast/TernaryExpr.java [hunk 4]
```diff
@@ -29,10 +29,13 @@
...
     TypeNode elseType = elseExpr().type();
-    if (thenType.equals(elseType)
-        || (thenType.isSupertypeOrEquals(elseType) && !thenType.equals(TypeNode.NULL))) {
+    if (thenType.isSupertypeOrEquals(elseType) && !thenType.equals(TypeNode.NULL)) {
       return thenType;
     }
+    if (thenType.equals(elseType)) {
+      // If types are boxed/primitive type equals, return the boxed type.
+      return TypeNode.isReferenceType(thenType) ? thenType : elseType;
+    }
```
Comment (xiaozhenliu-gg5) [inline src/main/java/com/google/api/generator/engine/ast/TernaryExpr.java:38]: add another condition based on the discussion, that when the types are boxed/primitive equal, we should the boxed type as the whole expression type.

File: src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java [hunk 5]
```diff
@@ -41,16 +38,13 @@ public void validTernaryExpr_primitiveType() {
             .setThenExpr(thenExpr)
             .setElseExpr(elseExpr)
             .build();
-    assertThat(ternaryExpr.conditionExpr().type()).isEqualTo(TypeNode.BOOLEAN);
-    assertThat(ternaryExpr.thenExpr().type()).isEqualTo(ternaryExpr.elseExpr().type());
-    assertThat(ternaryExpr.type()).isEqualTo(ternaryExpr.thenExpr().type());
+    assertTrue(ternaryExpr.conditionExpr().type().equals(TypeNode.BOOLEAN));
+    assertEquals(ternaryExpr.thenExpr().type(), ternaryExpr.elseExpr().type());
+    assertEquals(ternaryExpr.type(), TypeNode.INT);
   }
```
Comment (xiaozhenliu-gg5) [inline src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java:44]: the assert checks are updated to 
(1) the condition is boolean-typed  
(2) thenExpr/ elseExpr are equal/assignable to each other.
(3) the type of whole expression is correct.

File: src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java (hunk 1 above)
Comment (xiaozhenliu-gg5) [reply src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java:27]: This PR will be hold until #187 is merged.

File: src/main/java/com/google/api/generator/engine/ast/TernaryExpr.java (hunk 4 above)
Comment (miraleung) [reply src/main/java/com/google/api/generator/engine/ast/TernaryExpr.java:38]: Let's clean this up a bit more by separating them into a primitive-only check (that includes the boxed equality checking) and a purely-reference type check.

File: src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java [hunk 6]
```diff
@@ -67,6 +64,27 @@ public void validTernaryExpr_objectType() {
...
+        TernaryExpr.builder()
+            .setConditionExpr(conditionExpr)
+            .setThenExpr(thenExpr)
+            .setElseExpr(elseExpr)
+            .build();
+    assertEquals(ternaryExpr.conditionExpr().type(), TypeNode.BOOLEAN);
+    assertEquals(ternaryExpr.thenExpr().type(), ternaryExpr.elseExpr().type());
+    assertEquals(ternaryExpr.type(), TypeNode.INT_OBJECT);
+  }
+
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java:87]: Could you please confirm that we've covered all the edge cases in these tests?

File: src/main/java/com/google/api/generator/engine/ast/TernaryExpr.java (hunk 4 above)
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/TernaryExpr.java:42]: Sounds cool, first check `if (thenType.isPrimitiveType() || elseType.isPrimitiveType())` covers the condition either/both of types are primitive, then return the equal primitive type or the boxed type. Second check covers the reference types, we should always return the super /not-null type.

File: src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java (hunk 6 above)
Comment (xiaozhenliu-gg5) [reply src/test/java/com/google/api/generator/engine/ast/TernaryExprTest.java:89]: sure, added two valid unit tests that ensure:  the ternary expression has the correct type with primitive/boxed type expressions in different order. 
Also an invalid test to confirm the thenExpr/elseExpr should be primitive/boxed type equal.
